utils/fixed_decimal/src/decimal.rs

```diff
@@ -1300,32 +1300,220 @@ impl FixedDecimal {
+                // always equal or bigger than `i16::MIN`.
+            }
         }
-        let magnitude = position - 1;
-        self.upper_magnitude = cmp::max(self.upper_magnitude, magnitude);
 
-        // 2. If the rounding position is *lower than* the rightmost nonzero digit, exit early
-        if self.is_zero() || magnitude < self.nonzero_magnitude_end() {
+        // 2. If the rounding position is *lower than* the required digits to round to the next
+        // multiple, exit early.
+        let past_required_digits = match increment {
+            RoundingIncrement::MultiplesOf1 => {
+                // We can early exit if `position` is the last digit
+                position <= self.nonzero_magnitude_end()
+            }
+            RoundingIncrement::MultiplesOf2 | RoundingIncrement::MultiplesOf5 => {
+                // We can early exit if the digit at `position` is zero.
+                position < self.nonzero_magnitude_end()
+            }
+            RoundingIncrement::MultiplesOf25 => {
+                if position != i16::MAX {
+                    // We can early exit if the digits at `position` and `position + 1` are
+                    // both zeroes.
+                    position + 1 < self.nonzero_magnitude_end()
+                } else {
+                    false
+                }
+            }
+        };
+
+        if self.is_zero() || past_required_digits {
             #[cfg(debug_assertions)]
             self.check_invariants();
             return;
         }
 
         // 3. If the rounding position is *in the middle* of the nonzero digits
-        if magnitude < self.magnitude {
+        if position <= self.magnitude {
             // 3a. Calculate the number of digits to retain and remove the rest
-            let digits_to_retain = crate::ops::i16_abs_sub(self.magnitude, magnitude);
+            let digits_to_retain = crate::ops::i16_abs_sub(self.magnitude, position) + 1;
             self.digits.truncate(digits_to_retain as usize);
+
+            // 3b. Handle the last digit considering the increment.
+            let iter = match increment {
+                RoundingIncrement::MultiplesOf1 => {
+                    // Can just execute the algorithm normally.
+
+                    self.digits.iter_mut().rev().enumerate()
+                }
+                RoundingIncrement::MultiplesOf2 => {
+                    let mut iter = self.digits.iter_mut().rev().enumerate();
+
+                    // Must round the last digit to the next increment.
+                    let Some((_, digit)) = iter.next() else {
+                        debug_assert!(false, "`self.digits` should have at least a digit");
+                        return;
+                    };
+
+                    // Equivalent to (n + 2 / 2) * 2, which expands to the next
+                    // multiple of two
+                    *digit = (*digit + 2) & 0xFE;
```

Comment:
Note: When I first looked at this I questioned whether this should be `+1` instead of `+2` since only odd numbers are different between expand and trunc. I see now that `+2` is correct if there are additional trailing digits after the rounding position. However, I'm not completely sure that we handle cases where there are not trailing digits correctly. Can you check these cases?

| Input | Position | Increment | Expected Result |
|---|---|---|---|
| 0 | 0 | 2 | 0 |
| 0 | 0 | 5 | 0 |
| 0 | 0 | 25 | 0 |
| 0.1 | 0 | 2 | 2 |
| 0.1 | 0 | 5 | 5 |
| 0.1 | 0 | 25 | 25 |
| 1 | 0 | 2 | 2 |
| 1 | 0 | 5 | 5 |
| 1 | 0 | 25 | 25 |
| 2 | 0 | 2 | 2 |
| 2 | 0 | 5 | 5 |
| 2.1 | 0 | 2 | 4 |
| 2.1 | 0 | 5 | 5 |
| 4 | 0 | 2 | 4 |
| 4 | 0 | 5 | 5 |
| 4.1 | 0 | 2 | 6 |
| 4.1 | 0 | 5 | 5 |
| 5 | 0 | 2 | 6 |
| 5 | 0 | 5 | 5 |
| 5.1 | 0 | 2 | 6 |
| 5.1 | 0 | 5 | 10 |
| 6 | 0 | 2 | 6 |
| 6 | 0 | 5 | 10 |

Reply:
Thanks for the test cases! I think this caught that bug, and I also used the checks in `trunc_to_increment` to avoid doing more work if the last digits are already a multiple of the increment.